### PR TITLE
Microchip PolarFire SoC GPIO driver support

### PR DESCRIPTION
**Pull request by con-pax**
This patch adds driver support for Microchip's PolarFire SoC MSS GPIO.
Tested this patch on Microchip's PolarFire SoC ICICLE kit which has 3 Banks of GPIO's.
This patch also adds the GPIO node to mpfs-icicle.dts.

### CODE REVIEW

**Comment by con-pax** [comment]
Hi,

It looks like the Coding Guidelines CI check failed from a timeout?
Connecting to launchpad.net (launchpad.net)|2620:2d:4000:1001::8004|:443... failed: Network is unreachable.
can I request that it gets run again please?

File: boards/riscv/mpfs_icicle/mpfs_icicle_defconfig [hunk 1]
```diff
@@ -15,8 +15,10 @@ CONFIG_RISCV_SOC_INTERRUPT_INIT=y
...
 CONFIG_PLIC=y
 CONFIG_RISCV_MACHINE_TIMER=y
-CONFIG_GPIO=n
 CONFIG_XIP=n
 CONFIG_INIT_STACKS=y
 CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
 CONFIG_FPU=n
+# GPIO driver options
+CONFIG_GPIO=y
+CONFIG_GPIO_MCHP_MSS=y
```
Comment (galak) [inline boards/riscv/mpfs_icicle/mpfs_icicle_defconfig:24]: remove `CONFIG_GPIO_MCHP_MSS=y` should get handled automatically by Kconfig

File: drivers/gpio/Kconfig.mchp_mss [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Microchip PolarFire SoC Icicle Kit GPIO configuration options
+
+# Copyright (c) 2022 Microchip Technology Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config GPIO_MCHP_MSS
+	bool "Microchip PolarFire SoC GPIO driver"
+	depends on SOC_SERIES_RISCV64_MIV
```
Comment (galak) [inline drivers/gpio/Kconfig.mchp_mss:8]: remove `SOC_SERIES_RISCV64_MIV` and add:

```
default y
depends on DT_HAS_MICROCHIP_MPFS_GPIO_ENABLED
```

This utilizes an auto-generated Kconfig.dts file that will have the DT_HAS_MICROCHIP_MPFS_GPIO_ENABLED symbol defined.

File: dts/bindings/gpio/microchip,mpfs-gpio.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+compatible: "microchip,mpfs-gpio"
+
+include: [gpio-controller.yaml, base.yaml]
+
+properties:
+    reg:
+      required: true
+
+    label:
+      required: false
```
Comment (galak) [inline dts/bindings/gpio/microchip,mpfs-gpio.yaml:15]: remove `label`.

**Review by galak — DISMISSED** [review]
some kconfig & dts related changes needed.

File: boards/riscv/mpfs_icicle/mpfs_icicle_defconfig (hunk 1 above)
Comment (con-pax) [reply boards/riscv/mpfs_icicle/mpfs_icicle_defconfig:24]: Hi @galak,
Thanks for reviewing. Have finished testing requested changes locally and will push now.

File: drivers/gpio/Kconfig.mchp_mss (hunk 2 above)
Comment (con-pax) [reply drivers/gpio/Kconfig.mchp_mss]: done

File: dts/bindings/gpio/microchip,mpfs-gpio.yaml (hunk 3 above)
Comment (con-pax) [reply dts/bindings/gpio/microchip,mpfs-gpio.yaml]: done

File: drivers/gpio/gpio_mchp_mss.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+struct mss_gpio_t {
+	unsigned int gpio_cfg[32];
+	unsigned int gpio_irq;
+	const unsigned int gpio_in;
+	unsigned int gpio_out;
+	unsigned int gpio_cfg_all;
+	unsigned int gpio_cfg_byte[4];
+	unsigned int gpio_clr_bits;
+	unsigned int gpio_set_bits;
```
Comment (cfriedt) [inline drivers/gpio/gpio_mchp_mss.c:38]: It would be preferable to use uint32_t's here.

**Review by cfriedt — CHANGES_REQUESTED** [review]

File: dts/riscv/mpfs-icicle.dtsi [hunk 5]
```diff
@@ -92,5 +95,44 @@
...
+			ngpios = <32>;
+			status = "disabled";
+		};
+
+		gpio2: gpio@20122000 {
+			compatible = "microchip,mpfs-gpio";
+			reg = <0x20122000 0x1000>;
+			interrupt-parent = <&plic>;
+			interrupts = <53 1>;
+			//interrupt-controller;
```
Comment (fkokosinski) [inline dts/riscv/mpfs-icicle.dtsi:130]: Why is this commented out?

File: drivers/gpio/gpio_mchp_mss.c (hunk 4 above)
Comment (con-pax) [reply drivers/gpio/gpio_mchp_mss.c]: HI @cfriedt Thanks for the review, have made the changes and pushed.

File: dts/riscv/mpfs-icicle.dtsi (hunk 5 above)
Comment (con-pax) [reply dts/riscv/mpfs-icicle.dtsi]: hi @fkokosinski thanks for reviewing. this was my mistake. made the changes and pushed.

**Review by cfriedt — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Pull request merged**